### PR TITLE
JMAP mboxid_byrole hash

### PR DESCRIPTION
**Pull request by ksmurchison**
Introduce mboxid_byrole hash and use it for jmap_mailbox_find_role()

Note that is might be easier to view this by ignoring whitespace

### CODE REVIEW

**Review by rsto — CHANGES_REQUESTED** [review]
Looks good, but I'd prefer the code to be defined in jmap_api.c

File: imap/jmap_api.c [hunk 1]
```diff
@@ -435,6 +435,12 @@ HIDDEN void jmap_finireq(jmap_req_t *req)
 
     jmap_mbentry_cache_free(req);
 
+    if (req->mboxid_byrole) {
+        free_hash_table(req->mboxid_byrole, free);
+        free(req->mboxid_byrole);
```
[the remark this answers is not in the document]
Comment (ksmurchison) [reply imap/jmap_api.c]: done

File: imap/jmap_mailbox.c [hunk 2]
```diff
@@ -1920,14 +1945,13 @@ static char *_mbox_role_to_specialuse(const char *role)
 }
 
 EXPORTED int jmap_mailbox_find_role(jmap_req_t *req, const char *role,
```
[the remark this answers is not in the document]
Comment (ksmurchison) [reply imap/jmap_mailbox.c]: done

File: imap/jmap_mail.c [hunk 3]
```diff
@@ -13307,16 +13307,15 @@ static void _email_copy(jmap_req_t *req, json_t *copy_email,
     json_object_foreach_safe(jmailboxids, tmp, mbox_id, jval) {
         if (*mbox_id != '$') continue;
         const char *role = mbox_id + 1;
-        char *uniqueid = NULL;
-        if (jmap_mailbox_find_role(req, role, NULL, &uniqueid) == 0) {
+        const mbentry_t *mbentry = NULL;
+        if (jmap_mailbox_find_role(req, role, &mbentry) == 0) {
```
[the remark this answers is not in the document]
Comment (ksmurchison) [reply imap/jmap_mail.c]: done

**Review by rsto — APPROVED** [review]
LGTM

**Pull request merged**